v-next/hardhat-node-test-reporter/src/error-formatting.ts

```diff
@@ -108,17 +109,43 @@ function formatSingleError(
 ): string {
   const messageLines = [];
   const stackLines = [];
+  const nodeLines = [];
 
   for (const line of error.stack?.split("\n") ?? []) {
```

Comment:
It took me a while to understand why i found this harder to follow than what I was expecting, and I think I just got it.

If you were iterrating in reverse order, the algorithm is much simpler/closer to how we dicussed:

```ts
let foundNodeFrame = false;
let stackLines = [];

const lines = error.stack?.split("\n") ?? [];
for (let i = lines.length - 1; i >= 0; i--) {
  const line = lines[i];
  //...
  if (!isNodeLocation(reference.location) && !foundNodeFrame) {
    // We remove all the frames below the first non-node ones
    stackLines = [];
    foundNodeFrame = true; 
  }
   
  stackLines.push(formattedLine);
}

stackLines.reverse();
```

Note that this is pseudocode though, i haven't run it

Reply:
That's a fair point. It got a tad bit complicated with trying to extract lines from which we build up a message and discarding node lines, but not all of them, ...

This got me thinking and I put together another approach. I'm curious to see what you think about it.

First, I added a new interface [StackLine](https://github.com/NomicFoundation/hardhat/pull/5714/commits/bfe40154dd252547b04e67bd3b47f47d30ecadf8#diff-561a47ff066e39f4e07ec5fdc8265f35643511c1776154bd916966b8f2add17aR46) which can hold both the original line that we attempt to parse and a `StackReference` if we succeed to parse the line into it. With this we can [run all the lines through parseStackLine](https://github.com/NomicFoundation/hardhat/pull/5714/commits/bfe40154dd252547b04e67bd3b47f47d30ecadf8#diff-561a47ff066e39f4e07ec5fdc8265f35643511c1776154bd916966b8f2add17aR119) straight away and not worry wether we got undefined back.

Now, among all the parsed lines, we can find [the first one which has a defined stack reference](https://github.com/NomicFoundation/hardhat/pull/5714/commits/bfe40154dd252547b04e67bd3b47f47d30ecadf8#diff-561a47ff066e39f4e07ec5fdc8265f35643511c1776154bd916966b8f2add17aR122). With that in hand, we can [build a message](https://github.com/NomicFoundation/hardhat/pull/5714/commits/bfe40154dd252547b04e67bd3b47f47d30ecadf8#diff-561a47ff066e39f4e07ec5fdc8265f35643511c1776154bd916966b8f2add17aR132) from all the lines from the beginning up to that line.

We can also find [the last line which does **NOT** have a reference to node internals](https://github.com/NomicFoundation/hardhat/pull/5714/commits/bfe40154dd252547b04e67bd3b47f47d30ecadf8#diff-561a47ff066e39f4e07ec5fdc8265f35643511c1776154bd916966b8f2add17aR127). With that, we can easily [build the stack](https://github.com/NomicFoundation/hardhat/pull/5714/commits/bfe40154dd252547b04e67bd3b47f47d30ecadf8#diff-561a47ff066e39f4e07ec5fdc8265f35643511c1776154bd916966b8f2add17aR156) from all the lines between the first line with any stack reference and that line. With a special case being that the entire stack is node internals only.

Let me know what you think :)